src/dune_rules/pkg_build_progress.ml

```diff
@@ -0,0 +1,75 @@
+open! Import
+
+let enabled = Config.make_toggle ~name:"pkg_build_progress" ~default:`Disabled
+
+module Status = struct
+  type t =
+    [ `Downloading
+    | `Building
+    ]
+
+  let to_string = function
+    | `Downloading -> "Downloading"
+    | `Building -> "Building"
+  ;;
+end
+
+module Message = struct
+  type t =
+    { package_name : Package.Name.t
+    ; package_version : Package_version.t
+    ; status : Status.t
+    }
+
+  let user_message { package_name; package_version; status } =
+    let status_tag = User_message.Style.Ok in
+    User_message.make
+      [ Pp.concat
+          [ Pp.tag status_tag (Pp.text (Status.to_string status))
+          ; Pp.textf
+              " %s.%s"
+              (Package.Name.to_string package_name)
+              (Package_version.to_string package_version)
+          ]
+      ]
+  ;;
+
+  let display t =
+    match Config.get enabled with
+    | `Enabled -> Console.print_user_message (user_message t)
+    | `Disabled -> ()
+  ;;
+end
+
+module Spec = struct
+  type ('path, 'target) t =
+    { message : Message.t
+    ; output : Path.Build.t
+    }
+
+  let name = "progress-action"
+  let version = 1
+  let is_useful_to ~memoize:_ = true
+  let bimap { message; output } _f _g = { message; output }
+  let encode _ _ _ = Sexp.List []
+
+  let action t ~ectx:_ ~eenv:_ =
+    (match Path.Untracked.exists (Path.build t.output) with
+     | false -> Message.display t.message
+     | true -> ());
```

Comment:
Did you check if this clause ever executes? Isn't the target directory created every time a package is built?

Reply:
I now see why this can work. It's because the path in this action isn't being sandboxed.